File: MATLAB/SPCA&SCCA/SPCA/spca_amanpg.m
function [iter,F_amanpg, sparsity, time,X,Y_man]= spca_amanpg(B,option,F_palm)
% Alternating rgd and pgd for spca, i.e.
% if option.lambda < inf,  min Tr(Y'*B'B*Y)- 2Tr(X'*B'*B*Y) + lambda*norm(Y,'fro')^2  + mu*norm(Y,1) s.t. X'*X=In.
% if option.lambda = inf,  min - 2Tr(X'*B'*B*Y)+ norm(Y,'fro')^2  + mu*norm(Y,1) s.t. X'*X=In.
% If option.type =0; B is a m*d matrix.....m is sample size.
% d is the number of feature.  B should be centered and normalized such that
% the column means are 0 and the column Euclidean lengths are 1.
% Or option.type = 1, B is Gram matrix, i.e. covariance matrix ;
% data matrix is suppose to be normalized.
% e.g.
% if (type == 1) %covariance matrix
%     scale = max(diag(A)); % Sigma=A/scale;
% elseif (type == 0) %data matrix
%     A = A - repmat(mean(A,1),m,1);
%     scale = [];
%     for i = 1:d
%         scale = [scale norm(A(:,i))];
%     end
%     scale = max(scale);
%     A =A/scale;
%     %  Sigma=A'*A;
% end
tic;
%parameters
mu = option.mu; 
lambda = option.lambda; 
n = option.n; 
type = option.type;
[m,d] = size(B);
%h=@(X) mu*sum(sum(abs(X)));
h=@(X) sum(mu.*sum(abs(X)));
maxiter =option.maxiter;
tol = option.tol;
if d < m*2
    B = B'*B;
    type = 1;
end
if type == 0
    LY = 2*(svds(B,1))^2 + 2*lambda;
else
    LY = 2*svds(B,1) + 2*lambda;
end

%% initial point
X0 = option.X0;  Y0= option.Y0;
X = X0;  Y = Y0;
total_linesearch = 0;    linesearch_flag = 1;  min_step = 0;
linesearch_flag_Y = 1;
t = 1/LY;     tau = 100/d;
if type == 0
    AY = B'*(B*Y);    AX = B'*(B*X);
else
    AY = B*Y;     AX = B*X;
end

if lambda ~= inf %elastic net parameter isn't inf
    fx =  -2*sum(sum(X.*(AY))); 
    fy = sum(sum(Y.*(AY)))+lambda*norm(Y,'fro')^2  + h(Y);
    F_rgd(1)= fx + fy;
    for iter = 2:maxiter
 
        % update Y
        if linesearch_flag_Y == 0
            t = t*1.01;
        else
            t = max(1/LY, t/1.01);
        end
         linesearch_flag_Y = 0;
        Y_t = proximal_l1(Y-t*2*(AY-AX+lambda*Y), mu*t,n);
        if type == 0; AYt = B'*(B*Y_t);   else;  AYt = B*Y_t;     end
        f_ytrial = -2*sum(sum(X.*(AYt))) + sum(sum(Y_t.*AYt)) + lambda*norm(Y_t,'fro')^2 + h(Y_t);
        normpg = norm(Y_t - Y,'fro')^2/t^2;
        while f_ytrial > F_rgd(iter-1) - 1e-3*t* normpg
            t = 0.5*t;
            if t < 1e-5/d
                break;
            end
            Y_t = proximal_l1(Y-t*2*(AY-AX+lambda*Y), mu*t,n);
             if type == 0; AYt = B'*(B*Y_t);   else;  AYt = B*Y_t;     end
            f_ytrial = -2*sum(sum(X.*(AYt))) + sum(sum(Y_t.*AYt)) + lambda*norm(Y_t,'fro')^2 + h(Y_t);
            linesearch_flag_Y = 1;
           
        end
       Y = Y_t;  
       AY = AYt;
        % update X
        if linesearch_flag == 0
            tau = tau*1.1;
        end
        if min_step == 1
            tau = 1/d;
        end
        linesearch_flag = 0;
        min_step = 0;
        gx = -2*AY;   xgx = gx'*X;    
        %RGX = gx - X*xgx;  %canonical riemannian gradient
        RGX = gx-0.5*X*(xgx+xgx');%projected gradient
        TX = X - tau*RGX;  
        [U, SIGMA] = eig(TX'*TX);  
        SIGMA = diag(SIGMA);  J =U*diag(sqrt(1./SIGMA))*U';
        X_trial= TX*J;    
        f_xtrial = - 2*sum(sum(X_trial.*AY));
        fXval = - 2*sum(sum(X.*AY));
        normpg = norm(RGX,'fro')^2;
        while f_xtrial > fXval - 1e-3*tau* normpg
            tau = 0.5*tau;
            if tau < 1e-5/d
                min_step = 1;
                break;
            end
            TX = X - tau*RGX;
            %  [U,~,V]=svd(TX, 0); X_trial=U*V';
            %  [X_trial,R]=qr(TX,0);      X_trial =X_trial*diag(sign(diag(R)));
            [U, SIGMA] = eig(TX'*TX);  
            %[U, SIGMA] = eig(eye(n)+ tau^2*GtG);  
            SIGMA = diag(SIGMA);  J =U*diag(sqrt(1./SIGMA))*U';
            X_trial= TX*J;
            total_linesearch = total_linesearch+1;
            linesearch_flag = 1;
            f_xtrial = - 2*sum(sum(X_trial.*AY));
        end
        
        X = X_trial;
        %  AX = B'*(B*X);
        if type == 0; AX = B'*(B*X);   else;  AX = B*X;     end
        
        fx = f_xtrial;%- 2*sum(sum(X.*AY));
        fy = sum(sum(Y.*AY)) + lambda*norm(Y,'fro')^2 + h(Y);
        F_rgd(iter)= fx + fy;
        
        %  if  normDsquared<tol^2
        if iter >1
            if  ( abs(F_rgd(iter)-F_rgd(iter-1))<tol && F_rgd(iter) < F_palm) || abs(F_rgd(iter)-F_rgd(iter-1))< 1e-12
                break;
            end
        end
    end
else%elastic net parameter is inf
    fx =  -2*sum(sum(X.*(AY))); fy = norm(Y,'fro')^2  + h(Y);
    F_rgd(1)= fx + fy;
    for iter = 2:maxiter
        if linesearch_flag == 0
            tau = tau*1.1;
        end
        if min_step == 1
            tau = 1/d;
        end
        min_step = 0;
        linesearch_flag = 0;
        % update Y
        %Y_old = Y;
        t = 1/2;
        Y = proximal_l1(Y-2*t*(-AX+ Y), mu*t,n);
        if type == 0;  AY = B'*(B*Y);   else;  AY = B*Y;     end
        % update X
        gx = -2*AY;   xgx = gx'*X;     RGX = gx - X*xgx;  %canonical riemannian gradient
        TX = X - tau*RGX;
        [U,~,V]=svd(TX, 0);   X_trial=U*V';
        f_xtrial = - 2*sum(sum(X_trial.*AY));
        fXval = - 2*sum(sum(X.*AY));
        normpg = norm(RGX,'fro')^2;
        while f_xtrial > fXval - 1e-3*tau* normpg
            tau = 0.5*tau;
            if tau < 1e-3/d
                min_step = 1;
                break;
            end
            TX = X - tau*RGX;
            [U,~,V]=svd(TX, 0); X_trial=U*V';
            total_linesearch = total_linesearch+1;
            linesearch_flag = 1;
            f_xtrial = - 2*sum(sum(X_trial.*AY));
        end
        
        X = X_trial;
        %  [U,~,V]=svd(2*AY,0); X=U*V'; minimization
        %  AX = B'*(B*X);
        if type == 0; AX = B'*(B*X);   else;  AX = B*X;     end
        
        fx = f_xtrial;%- 2*sum(sum(X.*AY));
        fy =  norm(Y,'fro')^2 +h(Y);
        F_rgd(iter)= fx + fy;
        
        %  if  normDsquared<tol^2
        if iter >1
            if  abs(F_rgd(iter)-F_rgd(iter-1))<tol
                %  error(iter) =normpg +(norm(Y -Y_old,'fro')^2/t2^2);
                % if normpg + norm(Y -Y_old,'fro')^2/t^2< tol
                break;
            end
        end
    end
    
end
F_amanpg = F_rgd(iter);
Y_norm = sqrt(sum(Y.^2));
Y_norm(Y_norm == 0) = 1;
Y_man = Y./(ones(d,1)*Y_norm);
%X_man = X;
%Y_man = Y;
time = toc;
sparsity= sum(sum(Y==0))/(d*n);
%fprintf('%1.5f\n', tau);
%semilogy(1:iter, F_rgd - F_manpg);

File: R/AManPG/R/spca.amanpg.R
# if option.lambda < inf,  min Tr(Y'*B'B*Y)- 2Tr(X'*B'*B*Y) + lambda*norm(Y,'fro')^2  + mu*norm(Y,1) s.t. X'*X=In.
# if option.lambda = inf,  min - 2Tr(X'*B'*B*Y)+ norm(Y,'fro')^2  + mu*norm(Y,1) s.t. X'*X=In.
# TODO: Check if conjugate transpose is necessary. If not, get rid of Conj() calls.
# TODO: Consider custom gamma value?

spca.amanpg <- function(b, mu, lambda, n, x0, y0, f_palm, type=0, gamma=0.5,
                        maxiter=1e4, tol=1e-5, verbose=FALSE) {

  start <- Sys.time()

  dims <- dim(b);
  m <- dims[1];
  d <- dims[2];
  # anonymous function that gets sum of matrix X times mu
  h <- function(x) sum(mu * colSums(abs(x)))

  if (d < m * 2) {
    b <- Conj(t(b)) %*% b  # B^T B
    type <- 1
  }

  svds1 <- svd(b)$d[1]
  if (!type) ly <- 2 * svds1^2 + 2 * lambda else ly <- 2 * svds1 + 2 * lambda

  # initialization
  if (missing(x0) && missing(y0)) {  # assign initial point if not provided
    x0 <- svd(b, nv=n)$v
    y0 <- x0
  }

  x <- x0
  y <- y0
  total_linesearch <- 0
  linesearch_flag <- 1
  min_step <- 0
  linesearch_flag_y <- 1
  t <- 1 / ly
  tau <- 100 / d

  if (!type) {
    ay <- Conj(t(b)) %*% (b %*% y)
    ax <- Conj(t(b)) %*% (b %*% x)
  } else {
    ay <- b %*% y
    ax <- b %*% x
  }

  ### Main Loop ###
  if (is.finite(lambda)) {  # elastic net parameter isn't Inf
    fx <- -2 * sum(x * ay)
    fy <- sum(y * ay) + lambda * norm(y, 'F')^2 + h(y)
    f_rgd <- fx + fy  # will become vector in the loop

    for (iter in 2:maxiter) {
      if (verbose) {
        iter_start <- Sys.time()
        print("=========================")
        print(paste("On iteration", iter))
      }

      ### update y ###
      if (!linesearch_flag_y) t <- t * 1.01 else t <- max(1/ly, t/1.01)
      linesearch_flag_y <- 0

      y_t <- prox.l1(y - t*2 * (ay - ax + lambda * y), mu * t, n)

      if (!type) ayt <- Conj(t(b)) %*% (b %*% y_t) else ayt <- b %*% y_t
      f_ytrial <- -2 * sum(x * ayt) + sum(y_t * ayt) + lambda * norm(y_t, 'F')^2 + h(y_t)
      normpg <- norm(y_t - y, 'F')^2 / t^2

      # adjust step size to be appropriate
      while (f_ytrial > f_rgd[iter - 1] - 1e-3 * t * normpg) {
        t <- 0.5 * t  # gamma = 0.5
        if (t < 1e-5 / d) {
          break
        }

        y_t <- prox.l1(y - t * 2 * (ay - ax + lambda * y), mu * t, n)
        if (!type) ayt <- Conj(t(b)) %*% (b*y_t) else ayt <- b %*% y_t
        f_ytrial <- -2 * sum(x * ayt) + sum(y_t * ayt) + lambda * norm(y_t, 'F')^2 + h(y_t)
        linesearch_flag_y <- 1
      }

      y <- y_t
      ay <- ayt

      ### update x ###
      if (!linesearch_flag) tau <- tau*1.1
      if (min_step == 1) tau <- 1/d
      linesearch_flag <- 0
      min_step <- 0
      gx <- -2 * ay
      xgx <- Conj(t(gx)) %*% x
      # RGX <- gx - X %*% xgx  # Canonical Riemannian gradient
      rgx <- gx - 0.5 * x %*% (xgx + Conj(t(xgx)))  # Projected gradient
      tx <- x - tau * rgx

      eigendecomp <- eigen(Conj(t(tx)) %*% tx)
      u <- eigendecomp$vectors
      sigma <- eigendecomp$values
      j <- u %*% diag(sqrt(1 / sigma)) %*% Conj(t(u))
      x_trial <- tx %*% j
      f_xtrial <- -2 * sum(x_trial * ay)
      fxval <- -2 * sum(x * ay)
      normpg <- norm(rgx, 'F')^2

      while (f_xtrial > fxval - 1e-3 * tau * normpg) {
        tau <- 0.5 * tau
        if (tau < 1e-5 / d) {
          min_step <- 1
          break
        }

        tx <- x - tau * rgx
        eigendecomp <- eigen(Conj(t(tx)) %*% tx)
        u <- eigendecomp$vectors
        sigma <- eigendecomp$values
        j <- u %*% diag(sqrt(1 / sigma)) %*% Conj(t(u))  # note use of abs
        x_trial <- tx %*% j
        total_linesearch <- total_linesearch + 1
        linesearch_flag <- 1
        f_xtrial <- -2 * sum(x_trial * ay)
      }

      x <- x_trial
      if (!type) ax <- Conj(t(b)) %*% (b %*% x) else ax <- b %*% x
      fx <- f_xtrial
      fy <- sum(y * ay) + lambda * norm(y, 'F')^2 + h(y)
      f_rgd <- c(f_rgd, fx + fy)  # concatenate vector

      if (verbose) {
        print(paste("fx:", fx))
        print(paste("fy:", fy))
        print(paste("Finished with value", f_rgd[iter], "and difference", abs(f_rgd[iter]-f_rgd[iter-1])))
        print(difftime(Sys.time(), iter_start))
      }

      if (abs(f_rgd[iter] - f_rgd[iter - 1]) < tol && f_rgd[iter] < f_palm || abs(f_rgd[iter] - f_rgd[iter - 1]) < 1e-12) {
        if (verbose) print(paste("Difference of", abs(f_rgd[iter] - f_rgd[iter-1])))
        break
      }
    }
  } else {  # lambda is Inf
    fx <- -2 * sum(x * ay)
    fy <- norm(y, 'F')^2 + h(y)
    f_rgd <- fx + fy  # will become vector

    for (iter in 2:maxiter) {
      if (verbose) {
        iter_start <- Sys.time()
        print("=========================")
        print(paste("On iteration", iter))
      }

      if (!linesearch_flag) tau <- tau * 1.1
      if (min_step == 1) tau <- 1/d

      min_step <- 0
      linesearch_flag <- 0

      ### update y ###
      t <- 1/2
      y <- prox.l1(y - 2 * t * (-ax + y), mu * t, n)
      if (!type) ay <- Conj(t(b)) %*% (b %*% y) else ay <- b %*% y

      ### update X ###
      gx <- -2 * ay
      xgx <- Conj(t(gx)) %*% x
      rgx <- gx - x %*% xgx  # Canonical Riemannian gradient
      tx <- x - tau * rgx
      # tmp <- svd.econ(tx, type=0)
      tx.dims <- dim(tx)
      tmp <- svd(tx, nu=min(tx.dims[1], tx.dims[2]), nv=min(tx.dims[1], tx.dims[2]))
      x_trial <- tmp$u %*% Conj(t(tmp$v))
      f_xtrial <- -2 * sum(x_trial * ay)
      fxval <- -2 * sum(x * ay)
      normpg <- norm(rgx, 'F')^2

      while (f_xtrial > fxval - 1e-3 * tau * normpg) {
        tau <- 0.5 * tau
        if (tau < 1e-3 / d) {
          min_step <- 1
          break
        }

        tx <- x - tau * rgx
        # tmp <- svd.econ(tx, type=0)
        tmp <- svd(tx, nu=min(tx.dims[1], tx.dims[2]), nv=min(tx.dims[1], tx.dims[2]))
        x_trial <- tmp$u %*% Conj(t(tmp$v))
        total_linesearch <- total_linesearch + 1
        linesearch_flag <- 1
        f_xtrial <- -2 * sum(x_trial * ay)
      }

      x <- x_trial
      if (!type) ax <- Conj(t(b)) %*% (b %*% x) else ax <- b %*% x
      fx <- f_xtrial
      fy <- norm(y, 'F')^2 + h(y)
      f_rgd <- c(f_rgd, fx + fy)

      if (verbose) {
        print(paste("fx:", fx))
        print(paste("fy:", fy))
        print(paste("Finished with value", f_rgd[iter]))
        print(difftime(Sys.time(), iter_start))
      }

      # if normDsquared < tol^2, as in the e-stationary point

      if (abs(f_rgd[iter] - f_rgd[iter - 1]) < tol) {
        if (verbose) {
          print(paste("Final difference of", abs(f_rgd[iter] - f_rgd[iter - 1])))
        }
        break
      }
    }

  }

  ### Process return list ###
  y_norm <- sqrt(colSums(y^2))
  y_norm[y_norm == 0] <- 1

  results <- list(
    iter=iter,
    f_amanpg=f_rgd[iter],
    sparsity=sum(y == 0) / (d * n),
    time=difftime(Sys.time(), start),  # TODO: Check if acceptable value of time
    x=x,
    y_man=y / (matrix(1, d, 1) %*% y_norm)
  )

  return(results)
}
